### PR TITLE
Fix broken nix-build on darwin add vim related nix packages

### PR DESCRIPTION
**Pull request by countoren**
realtes to this issue [Nix: add Vim plugins #144](https://github.com/quick-lint/quick-lint-js/issues/144)

### CODE REVIEW

File: dist/nix/quick-lint-js.nix [hunk 1]
```diff
@@ -15,15 +14,25 @@ mkDerivation {
...
+      filter = name: type: !(lib.hasSuffix ".nix" name);
+      src = ../../.;
+    };
+  };
   unpackPhase = null;
+  buildInputs = [ pkgs.python310 ];
 
-  nativeBuildInputs = with pkgs; [ cmake ninja ];
+  nativeBuildInputs = with pkgs; [
+    clang_11 #does not work with clang_12 or clang_7
```
Comment (strager) [inline dist/nix/quick-lint-js.nix:28]: Shouldn't this dependency on `clang_11` only exist on Darwin (macOS), not on Linux?

File: dist/nix/quick-lint-js.nix [hunk 2]
```diff
@@ -1,9 +1,8 @@
 # Copyright (C) 2020  Matthew "strager" Glazar
 # See end of file for extended copyright information.
 
-{ pkgs
+{ pkgs ? import <nixpkgs>{}
```
Comment (strager) [inline dist/nix/quick-lint-js.nix:4]: What's this change for? If it's not related to Python or Clang or caching, put it in its own commit with an explanation.

File: dist/nix/quick-lint-js.nix [hunk 3]
```diff
@@ -15,15 +14,25 @@ mkDerivation {
   pname = "quick-lint-js";
   version = builtins.head (builtins.split "\n" (builtins.readFile ../../version));
 
-  src = ../../.;
+  src = with pkgs; lib.cleanSourceWith {
+    filter = lib.cleanSourceFilter;
+    src = lib.cleanSourceWith {
+      filter = name: type: !(lib.hasSuffix ".nix" name);
```
Comment (strager) [inline dist/nix/quick-lint-js.nix:20]: Why are we ignoring .nix files?

File: dist/nix/vim-with-quick-lint-js.nix [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{ pkgs ? import <nixpkgs> {} 
+, vimrcWithPlugins ? import ./vimrc-with-plugins.nix { inherit pkgs; } 
+, vim  ? pkgs.vim #Pick your vim
+, name ? "jsvim"
+}:
+pkgs.symlinkJoin {
+  inherit name;
+  paths = [ (pkgs.vim_configurable.override { python = pkgs.python3; }) ];
```
Comment (strager) [inline dist/nix/vim-with-quick-lint-js.nix:8]: Why do we need Python 3 support in Vim?

File: dist/nix/vim-with-quick-lint-js.nix [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{ pkgs ? import <nixpkgs> {} 
+, vimrcWithPlugins ? import ./vimrc-with-plugins.nix { inherit pkgs; } 
+, vim  ? pkgs.vim #Pick your vim
+, name ? "jsvim"
```
Comment (strager) [inline dist/nix/vim-with-quick-lint-js.nix:4]: Call it quick-lint-js-vim or vim-quick-lint-js or something.

File: dist/nix/vimrc-with-plugins.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
+{ pkgs ? import <nixpkgs>{}
+, additionalPlugins? []
+, additionalCustPlugins? {}
+, additionalVimrc? ""#(pkgs.builtins.readFile ./vimrc)
```
Comment (strager) [inline dist/nix/vimrc-with-plugins.nix:4]: Remove debug comment?
```suggestion
, additionalVimrc? ""
```

File: dist/nix/quick-lint-js.nix (hunk 1 above)
Comment (strager) [inline dist/nix/quick-lint-js.nix:28]: `clang_11` isn't needed at run-time, so it should be in `buildInputs`, not `nativeBuildInputs`.

File: dist/nix/vimrc-with-plugins.nix [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+}:
+with pkgs;
+let quick-lint-js = import ./quick-lint-js.nix { inherit pkgs;} ;
+in
+vimUtils.vimrcFile {
+  customRC =
+    ''
+      let g:ale_javascript_quick_lint_js_executable = '${quick-lint-js}/bin/quick-lint-js'
+    ''+additionalVimrc;
```
Comment (strager) [inline dist/nix/vimrc-with-plugins.nix:13]: Instead of inserting a line into the vimrc, we could generate a `plugin/quick-lint-js-nix.vim` file in the `buildPhase` of the plugin derivation (line 15).

File: dist/nix/vimrc-with-plugins.nix [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+   vam.knownPlugins = vimPlugins // additionalCustPlugins // {
+     quick-lint-js = vimUtils.buildVimPluginFrom2Nix {
+       name = "quick-lint-js";
+       src = "${quick-lint-js}/share/vim/vimfiles";
+     };
```
Comment (strager) [inline dist/nix/vimrc-with-plugins.nix:18]: Could we have this part express the dependency on ALE? Maybe something like this (based on `vim-addon-xdebug` in `<nixpkgs>/pkgs/misc/vim-plugins/overrides.nix`):

    dependencies = with self; [ ale ];

**Review by strager — COMMENTED** [review]
Instead of creating a vimrc file, we should just export the `vimUtils.buildVimPluginFrom2Nix` part. Then users can list `"quick-lint-js"` in their own `vam.pluginDictionaries`. (I think? I don't know. I never used Nix to manage this stuff.) (We can still provide a derivation for Vim which uses the plugin.)

File: dist/nix/quick-lint-js.nix (hunk 1 above)
Comment (countoren) [reply dist/nix/quick-lint-js.nix:28]: nix will bring clang_11 in both case and will use it only during build

File: dist/nix/vim-with-quick-lint-js.nix (hunk 4 above)
Comment (countoren) [reply dist/nix/vim-with-quick-lint-js.nix:8]: the idea of this file is to be an example file, you we dont need it.

File: dist/nix/vimrc-with-plugins.nix (hunk 6 above)
Comment (countoren) [reply dist/nix/vimrc-with-plugins.nix:4]: it meant to be an example line if you want to add your own vimrc configuration. you could use this function either passing your vimrc this way or change this default value and copy paste it.

File: dist/nix/quick-lint-js.nix (hunk 1 above)
Comment (countoren) [reply dist/nix/quick-lint-js.nix:28]: in order to work clang_11 had to be in ```nativeBuildInputs``` for me

Comment (strager) [reply dist/nix/quick-lint-js.nix:28]: That seems wasteful on Linux where GCC is the default. If we don't have a good reason to use Clang on Linux, we should avoid Clang on Linux.

In my experience, GCC generates faster executables than Clang does. We should prefer GCC if it works and is the default anyway.
